Implement Delete, Create File, Create Folder for the right click menu in ProjectNavigator

KaiTheRedNinja opened the pull request:
# Description

TL;DR, the right-click menu in the Project Navigator now allows you to create and delete files and folders. The WorkspaceClient has been updated to detect all file system changes and handle them correctly.

## Very detailed list of changes:
- CodeEdit/NavigatorSidebar/ProjectNavigator/OutlineView/OutlineMenu.swift
  - `openInTab`: Opens the file similar to double-clicking it
  - `openInExternalEditor`: Opens the file in the default editor as if it were clicked in finder
  - `newFile` and `newFolder`: Creates a new file/folder at the given location named "Untitled" (with a number, if the file already exists with that name)
- CodeEdit/NavigatorSidebar/ProjectNavigator/OutlineView/OutlineViewController.swift
  - Passes the `reloadData` function to the `WorkspaceClient` for it to call when the file system updates
  - Passes the `workspace` variable to the `OutlineMenu` so that it can ask it to create files or folders
- CodeEditModules/Modules/WorkspaceClient/src/Interface.swift
  - Defines `onRefresh`, a variable holding a function. It is set by the `OutlineViewController`, and run when a file system change has been detected and handled. 
- CodeEditModules/Modules/WorkspaceClient/src/Live.swift
  - `rebuildFiles`: Given a file item, it deletes its children or creates new ones so that it is accurate to the file system. This is a recursive function, so it rebuilds an entire directory's contents and subcontents, not just the immediate ones.
  - `startListeningToDirectory`: Edited so that it uses multiple dispatch sources instead of just one. I found that the previous (single) listener only detected changes at the top level. This new function applies a dispatch source to every directory to make sure that every file change is detected.
  - `stopListeningToDirecotry`: Edited to close and delete every dispatch source
- CodeEditModules/Modules/WorkspaceClient/src/Model/FileItem.swift
  - `openWithExternalEditor`: Opens the file in the default editor as if it were clicked in finder. Called by the OutlineMenu.
  - `addFolder` and `addFile`: Edited so that when a file with the same name exists, a number is appended to the back until that new name is unused.
  - `delete`: Edited so that it deletes the contents of the directory first and then the directory (async so that its faster), to avoid a bad access error.

# Related Issue

* #351
I'm sorry if I couldn't find a related issue, there didn't seem to be one specifically related to creating or deleting files.

# Checklist

Code todo:
- [x] Don't recreate Dispatch File Sources every time a file system change is detected
- [x] Fix bug that causes large folders to create ERROR BAD ACCESS errors when deleted (this is because the item is deleted while the indexing code is running, so suddenly the item has been de-init'd and causes the error)

Pull Request todo:
- [x] I read and understood the [contributing guide](https://github.com/CodeEditApp/CodeEdit/blob/main/CONTRIBUTING.md) as well as the [code of conduct](https://github.com/CodeEditApp/CodeEdit/blob/main/CODE_OF_CONDUCT.md)
- [x] My changes generate no new warnings
- [x] My code builds and runs on my machine
- [x] I documented my code
- [ ] Review requested

# Screenshots

Videos of different permutations of creating and deleting files within folders:

https://user-images.githubusercontent.com/88234730/172625011-3ba50b6d-1ab2-4acc-b2c4-0faa934946b7.mov

https://user-images.githubusercontent.com/88234730/172625061-9762edd2-2e79-4d21-b852-ceddb9ade6ac.mov

https://user-images.githubusercontent.com/88234730/172625069-8b3340ec-4af7-4685-bfaf-8fc52cc2f0f7.mov

https://user-images.githubusercontent.com/88234730/172625071-90a29041-4a9a-4c70-b746-98e284f73c62.mov

wdg commented on CodeEditModules/Modules/WorkspaceClient/src/Live.swift:
```diff
@@ -54,41 +58,141 @@ public extension WorkspaceClient {
+            // iterate over every item, checking if its a directory first
+            for item in flattenedFileItems.values {
+                // check if it actually exists, doesn't have a listener, and is a folder
+                guard item.isFolder else { continue }
+                guard !sources.keys.contains(item.id) else { continue }
+                guard FileItem.fileManger.fileExists(atPath: item.url.path) else { continue }
```
Since you are not throwing any errors, i think you can better chain this guard.

```swift
guard item.isFolde,
           !sources.keys.contains(item.id)
           FileItem.fileManger.fileExists(atPath: item.url.path) else { continue }
```

wdg reviewed (approved):
LGTM

KaiTheRedNinja commented:
Fixed bug that causes EXC_BAD_ERROR when folders with many files are deleted, plus addressing the points raised by wdg. 

CodeEditModules/Modules/WorkspaceClient/src/Live.swift:
- `rebuildFiles` function now returns a boolean representing if there were any changes made
  - `newChildren` variable removed as it was redundant
- `startListeningToDirectory` function edited so that only one instance of the indexer is running at any given time
  - New variables:
    - `isRunning`: Boolean that dictates if an instance of the indexer is running.
    - `anotherInstanceRan`: If an instance of the indexer is running and another file update is detected, it increases `anotherInstanceRan` instead of running another concurrent indexer and causing illegal memory access. When the indexer has stopped running, it checks if `anotherInstanceRan` is > 0, if so it re-indexes the files again.

lukepistrol commented:
Please make sure to mark this PR `Ready for Review` once you're done

lukepistrol commented on CodeEdit/NavigatorSidebar/NavigatorSidebarToolbarBottom.swift:
```diff
@@ -34,8 +34,16 @@ struct NavigatorSidebarToolbarBottom: View {
+                guard let root = try? workspace.workspaceClient?.getFileItem(folderURL.path) else { return }
+                root.addFolder(folderName: "untitled") // TODO: use currently selected file instead of root
+            }
             Button("Not implemented yet") {}
```
I guess this can be deleted now

lukepistrol commented on CodeEdit/NavigatorSidebar/NavigatorSidebarToolbarBottom.swift:
```diff
@@ -34,8 +34,16 @@ struct NavigatorSidebarToolbarBottom: View {
         Menu {
-            Button("Add File") {}
-                .disabled(true)
+            Button("Add File") {
+                guard let folderURL = workspace.workspaceClient?.folderURL() else { return }
+                guard let root = try? workspace.workspaceClient?.getFileItem(folderURL.path) else { return }
+                root.addFile(fileName: "untitled") // TODO: use currently selected file instead of root
+            }
+            Button("Add Folder") {
+                guard let folderURL = workspace.workspaceClient?.folderURL() else { return }
+                guard let root = try? workspace.workspaceClient?.getFileItem(folderURL.path) else { return }
+                root.addFolder(folderName: "untitled") // TODO: use currently selected file instead of root
+            }
```
Crashes on projects located on Desktop sometimes.

> Thread 1: EXC_BAD_ACCESS (code=1, address=0x3ff0000000000008)

KaiTheRedNinja replied:
What do you mean "crashes on projects"? Do you mean when you click the right click menu items? Also, if you recreate the exact setup in another folder, does it work? I've always used a folder on my desktop to test it out and it worked.

KaiTheRedNinja replied:
@lukepistrol did you have the same issue that wdg had with the addFile function not being updated with my changes?

wdg replied:
I have reopened https://github.com/CodeEditApp/CodeEdit/issues/628, so that we can make progress on merging this PR.

lukepistrol commented on CodeEdit/NavigatorSidebar/ProjectNavigator/OutlineView/OutlineMenu.swift:
```diff
@@ -169,11 +177,45 @@ final class OutlineMenu: NSMenu {
+        item?.openWithExternalEditor()
+    }
+
+    // TODO: allow custom file names
+    /// Action that creates a new untitled file
+    @objc
+    private func newFile() {
+        item?.addFile(fileName: "untitled")
+    }
```
If a file already exists it will crash with:

> Thread 1: Fatal error: Duplicate keys of type 'FileItem' were found in a Dictionary.

KaiTheRedNinja replied:
In [CodeEditModules/Modules/WorkspaceClient/src/Model/FileItem.swift](https://github.com/CodeEditApp/CodeEdit/pull/655/files#diff-36567e2ccad007c9513fbeaf771729619fe2ab185cebff08fe8a467a16b1a984) it does check for duplicates? This error should not be happening. Can you check if the addFile function has been changed on your end? It should look like
```swift
public func addFile(fileName: String) {
    let fileUrl = url.appendingPathComponent(fileName)
    // check if folder, if it is create file under self
    var fileUrl = (self.isFolder ?
        self.url.appendingPathComponent(fileName) :
        self.url.deletingLastPathComponent().appendingPathComponent(fileName))

    // if a file/folder with the same name exists, add a number to the end.
    var fileNumber = 0
    while FileItem.fileManger.fileExists(atPath: fileUrl.path) {
        fileNumber += 1
        fileUrl = fileUrl.deletingLastPathComponent().appendingPathComponent("\(fileName)\(fileNumber)")
    }

    // create the file
    FileItem.fileManger.createFile(
        atPath: fileUrl.path,
        contents: nil,
        attributes: [FileAttributeKey.creationDate: Date()]
    )
}
```

lukepistrol replied:
looks the same. Pulled it straight from your branch.

KaiTheRedNinja replied:
This is actually an error with the new editor. If you close any open editors, the error does not appear.

wdg reviewed (changes requested):
Code looks good, please check for the crashes what luke described.
~i think if you check at `newFile()` if the file already exists, that you can fix the crashes.~

~`addFile()` is not checking for duplicates.~
Edit: Removed code display of main branch.

KaiTheRedNinja commented:
I'm a bit puzzled by the issues you're having with the addFile function. At [CodeEditModules/Modules/WorkspaceClient/src/Model/FileItem.swift](https://github.com/CodeEditApp/CodeEdit/pull/655/files#diff-36567e2ccad007c9513fbeaf771729619fe2ab185cebff08fe8a467a16b1a984), the function does actually check for duplicate file items.

wdg commented:
> I'm a bit puzzled by the issues you're having with the addFile function. At [CodeEditModules/Modules/WorkspaceClient/src/Model/FileItem.swift](https://github.com/CodeEditApp/CodeEdit/pull/655/files#diff-36567e2ccad007c9513fbeaf771729619fe2ab185cebff08fe8a467a16b1a984), the function does actually check for duplicate file items.
> 
> @lukepistrol @wdg do your addFile functions look like
> 
> ```swift
> public func addFile(fileName: String) {
>     let fileUrl = url.appendingPathComponent(fileName)
>     // check if folder, if it is create file under self
>     var fileUrl = (self.isFolder ?
>         self.url.appendingPathComponent(fileName) :
>         self.url.deletingLastPathComponent().appendingPathComponent(fileName))
> 
>     // if a file/folder with the same name exists, add a number to the end.
>     var fileNumber = 0
>     while FileItem.fileManger.fileExists(atPath: fileUrl.path) {
>         fileNumber += 1
>         fileUrl = fileUrl.deletingLastPathComponent().appendingPathComponent("\(fileName)\(fileNumber)")
>     }
> 
>     // create the file
>     FileItem.fileManger.createFile(
>         atPath: fileUrl.path,
>         contents: nil,
>         attributes: [FileAttributeKey.creationDate: Date()]
>     )
> }
> ```
> 
> Because @wdg quoted a file from the main repo, not my branch, and if the functions are there then there should be no crashes. I've tested same-file names to death and you can see them in action in the videos at the initial comment.

Oops, when i searched it searched on the main branch.
https://github.com/CodeEditApp/CodeEdit/blob/23576f7ae7c771be394954f2c118c1d66a5ba139/CodeEditModules/Modules/WorkspaceClient/src/Model/FileItem.swift#L187-L208

I tried to search why https://github.com/CodeEditApp/CodeEdit/pull/655#discussion_r895418586 is happening.

austincondiff commented:
When creating a new file, it needs to be visible (parent folder expanded) and the text needs to be selected so that the user can rename it from untitled.

KaiTheRedNinja commented:
> When creating a new file, the new file needs to be visible (parent folder expanded) and the text needs to be selected so that the user can rename it from "untitled".

The former can be done, but FYI currently there is no feature (that I can find) that allows for the renaming of files.

KaiTheRedNinja commented:
@lukepistrol it turns out that the duplicate keys error and EXC BAD ACCESS aren't actually from my code, it's from the new editor's and it is triggered whether the file change happened inside CodeEdit (via the right-click menu) or outside (via finder). 

@austincondiff files can now be renamed, and there is also file name validation (when its invalid, the text box turns mild red)

lukepistrol commented:
This is not a problem with the new editor but with either CodeFile or WorkspaceDocument. The editor only interfaces with each CodeFile's content string as a Binding. It appears to happen when a saving operation forces the CodeFile to be reloaded from disk which already happened before using the new editor

lukepistrol reviewed (approved):
LGTM 👍

austincondiff commented:
@allcontributors add @KaiTheRedNinja for code
